pyk/src/pyk/konvert/_kast_to_kore.py

```diff
@@ -198,45 +218,105 @@ def _inject(definition: KDefinition, term: KInner, sort: KSort) -> KInner:
 
 # 'krule' should have sorts on variables
 def krule_to_kore(definition: KDefinition, krule: KRule) -> Axiom:
+
     krule_body = krule.body
     krule_lhs_config = extract_lhs(krule_body)
     krule_rhs_config = extract_rhs(krule_body)
-    krule_lhs_constraints = [bool_to_ml_pred(c) for c in flatten_label('_andBool_', krule.requires) if not c == TRUE]
-    krule_rhs_constraints = [bool_to_ml_pred(c) for c in flatten_label('_andBool_', krule.ensures) if not c == TRUE]
-    krule_lhs = mlAnd([krule_lhs_config] + krule_lhs_constraints)
-    krule_rhs = mlAnd([krule_rhs_config] + krule_rhs_constraints)
-
-    top_level_kore_sort = SortApp('SortGeneratedTopCell')
-    top_level_k_sort = KSort('GeneratedTopCell')
-    # The backend does not like rewrite rules without a precondition
-    if len(krule_lhs_constraints) > 0:
-        kore_lhs: Pattern = kast_to_kore(definition, krule_lhs, sort=top_level_k_sort)
+
+    is_functional = isinstance(krule_lhs_config, KApply) and krule_lhs_config.label.name in definition.function_labels
+
+    top_level_k_sort = KSort('GeneratedTopCell') if not is_functional else definition.sort_strict(krule_lhs_config)
+    top_level_kore_sort = _ksort_to_kore(top_level_k_sort)
+
+    # Do sort inference on the entire rule at once
+    kast_lhs = mlAnd(
+        [krule_lhs_config]
+        + [
+            bool_to_ml_pred(constraint, sort=top_level_k_sort)
+            for constraint in flatten_label('_andBool_', krule.requires)
+            if not constraint == TRUE
+        ],
+        sort=top_level_k_sort,
+    )
+    kast_rhs = mlAnd(
+        [krule_rhs_config]
+        + [
+            bool_to_ml_pred(constraint, sort=top_level_k_sort)
+            for constraint in flatten_label('_andBool_', krule.ensures)
+            if not constraint == TRUE
+        ],
+        sort=top_level_k_sort,
+    )
+    kast_rule_sorted = definition.sort_vars(KRewrite(kast_lhs, kast_rhs))
+
+    kast_lhs_body, *kast_lhs_constraints = flatten_label('#And', extract_lhs(kast_rule_sorted))
+    kast_rhs_body, *kast_rhs_constraints = flatten_label('#And', extract_rhs(kast_rule_sorted))
+    kore_lhs_body = kast_to_kore(definition, kast_lhs_body, sort=top_level_k_sort)
+    kore_rhs_body = kast_to_kore(definition, kast_rhs_body, sort=top_level_k_sort)
+
+    axiom_vars: tuple[SortVar, ...] = ()
+    kore_axiom: Pattern
+    if not is_functional:
```

Comment:
Consider extracting two helpers: one for the functional case, one for a normal rewrite.

Reply:
I think I'll leave it as-is for now. But I can make an issue with the desired refactorings listed there.

Reply:
https://github.com/runtimeverification/k/issues/4690